src/lib/distances.jl

```diff
@@ -1,22 +1,23 @@
 using .Distances
+import .ChainRules: NO_FIELDS, rrule
 
-@adjoint function (::SqEuclidean)(x::AbstractVector, y::AbstractVector)
+function rrule(::SqEuclidean, x::AbstractVector, y::AbstractVector)
```

Comment:
I think it's not a good idea to add these type piracies to Zygote. I assume it is better to move it to e.g. ChainRules or Distances (IIRC the main developers did not want to add rules...) right away. Otherwise I fear it will be not only surprising that Zygote adds these rules (and one has to load Zygote to get these chainrules definitions) but also more annoying if one moves it to another package later on (since one has to be careful to avoid method redefinitions etc.).

A temporary workaround could also be to define the rules in Zygote, but only for `ZygoteRuleConfig`.

Reply:
Sure, but the problem is that if I start to add these rules in ChainRules, will we be able to test them?

Reply:
In ChainRules? There you would just test them with ChainRulesTestUtils. Actually, this should be done regardless of where they are defined to ensure that the rules are implemented properly, return the correct types, have the correct signature, are type stable etc.

Reply:
Xref https://github.com/JuliaArrays/FillArrays.jl/pull/153 maybe. I don't think CR.jl wants to provide rules for, and hence depend on, other packages. If they don't want to depend on CRC.jl, then the other option is to make tiny glue packages for the rules. Or to wait for Pkg to build in this ability. 

But setting up rules using ZygoteRuleConfig here can't hurt, and would make it easier to move them elsewhere later.

Reply:
Would ChainRules be willing to take on Distances.jl as a dependency? I assumed the only path for an external package outside of upstreaming to Distances itself would be something like DistributionsAD.jl.

In the short-medium term interest of removing as many (ideally all) `adjoint`s from Zygote itself, using `ZygoteRuleConfig` is a great option.

Reply:
Please not something like DistributionsAD :smile: It's a complete mess, full of type piracy, and to a large extent it contains workarounds and hacks that should have been added to the respective AD backends or other upstream packages such as Distributions or PDMats. The long term plan is to get rid of it as much as possible, and already all CR things were moved to Distributions. While it speeded up initial development to accumulate all the stuff in the package instead of fixing other packages, we now have to pay the price for it and spend a significant amount of time with cleaning it.

Reply:
Not saying it's a good idea :P, more that it's the only option outside of hosting rules in the AD backend if upstream is not willing to accept PRs.

Reply:
:stuck_out_tongue: 

IMO defining the rules for ZygoteRuleConfig only is probably the easiest and cleanest option if one wants to avoid (repeating) these discussions about whether it rather belongs to ChainRules/Distances/etc. here.

Reply:
>Would ChainRules be willing to take on Distances.jl as a dependency?

The policy is that ChainRules provides rules for functionality in Julia Base. Packages should depend on ChainRulesCore (which is super light and has no dependencies other than Julia Base) to define rules for their functionality. @oxinabox would have to approve exceptions, but I don't think I can remember any, apart from two in the early days that have since been moved out.

I agree that the use of ZygoteRuleConfig is the best solution for now. If we decide in the future where we want to put them it would be made very easy as well.